### PR TITLE
Benchmarking vignette

### PR DESCRIPTION
**Pull request by schloerke**
Hi @dselivanov !

From request: https://github.com/the-benchmarker/web-frameworks/pull/3334#issuecomment-700169246

I've updated the document to try to have plumber do the same code style path as what RestRServe performs. I added in `plumber + future` into the comparison image at the bottom.

Please re-run the document chunks manually to regenerate the images using your machine (`plot_results()` calls `ggsave()`).

Other updates:
* We are not trying to benchmark json conversion, so I have each app returning `text/plain`.
* `calc_fib(15)`, `calc_fib(20)`, and `calc_fib(25)` had their median times scale by a factor of 10x on my machine. I figured this was a good comparison of "low", "medium", and "high" computation
* I adjusted the y scale to use the expansion parameter, rather than using a fixed addition. (Would not work well with the comparison plot.)
* The log of the background process is printed after the process has completed. This helps with debugging and making sure each process ran as expected.

I am shocked as to how good future works within plumber. I was not expecting that. The performance could be dramatically reduce if the amount of communication was increased (which RestRServe would not need to worry about).

This is your document. Please update/adjust/tweak/toss/include any of the content.

Best,
Barret

### CODE REVIEW

**Comment by schloerke** [comment]
btw... from my local testing of other frameworks in https://github.com/the-benchmarker/web-frameworks, I think R as a whole is going to get crushed. 😞

**Comment by dselivanov** [comment]
@schloerke thanks for PR! It will take some time to review (hopefully early next week).

File: vignettes/benchmarks/Benchmarks.Rmd [hunk 1]
```diff
@@ -156,69 +247,94 @@ maintain pool of connections to RestRserve and hence won't suffer from creating
...
-  ylab("requests per second") +
-  theme_minimal() +
-  theme(plot.title = element_text(hjust = 0.5)) + 
-  scale_fill_manual(values = colour_pal) +
-  scale_color_manual(values = colour_pal) +
-  labs(fill = "package", col = NULL) +
-  ggtitle("plumber vs RestRserve")
-```
-<img src="../img/bench-rps-vs-plumber.png" width="640"/>
+Due to the overhead of creating a new process for each request in `RServe`, `RestRserve` does not perform as quickly as `plumber` when computing instantaneous routes.  However, `RestRserve` shows it's strength when executing routes that have both low and high computational costs. No extra coding logic is needed to leverage `RestRserve`'s multi-threaded execution. Mixing `plumber` and `future` together for _high computation_ routes brings performance that scales with the number of threads (similar to `RestRserve`), but at the cost of extra route logic and domain knowledge.
```
Comment (schloerke) [inline vignettes/benchmarks/Benchmarks.Rmd:340]:
```suggestion
Due to the overhead of creating a new process for each request in `RServe`, `RestRserve` does not perform as quickly as `plumber` when computing instantaneous routes.  However, `RestRserve` shows it's strength when executing routes that have both low and high computational costs. No extra coding logic is needed to leverage `RestRserve`'s multi-threaded execution. Mixing `plumber` and `future` together for _high computation_ routes brings performance that scales with the number of concurrent requests (similar to `RestRserve`), but at the cost of extra route logic and domain knowledge.
```

File: vignettes/benchmarks/Benchmarks.Rmd [hunk 2]
```diff
@@ -69,85 +85,160 @@ Configurations:
...
     scale_color_manual(values = colour_pal) +
-    labs(fill = "Runtime Checks", col = NULL) +
-    ggtitle(title)
+    theme_minimal() +
+    theme(
+      plot.title = element_text(hjust = 0.5),
+      legend.position = "bottom"
+    ) + 
+    labs(
+      x = "number of threads",
```
Comment (schloerke) [inline vignettes/benchmarks/Benchmarks.Rmd:189]:
```suggestion
      x = "concurrent requests",
```

**Comment by schloerke** [comment]
Had feedback to update the axis labels to avoid confusion. The plots will need to be re-rendered.

**Comment by dselivanov** [comment]
I'm really sorry that it takes so long to review. Very busy last weeks. Mb @artemklevtsov can help with initial review?

**Comment by schloerke** [comment]
@dselivanov Take your time. I understand. Hopefully it can be of use before your next release.

**Comment by dselivanov** [comment]
@schloerke sorery once again that it took that long. I've updated couple things and pushed in your branch. Major is `keep-alive` parameter (which was 10000 and now -1). It seems didn't work properly on newer versions on apib (mb it is in nanoseconds?).

`future` seems introduce significant latency and it doesn't look it worth to use it when computation requirement is moderate...

**Comment by dselivanov** [comment]
Also I've switched to desktop machine for benchmarks. On laptop they are not reliable at all - system changes CPU frequency very aggressively depending on temperature and number if cores used (each time I run results were considerably different).

**Comment by dselivanov** [comment]
Another observation is that `plumber` with keep alive disabled performs better compared to keep alive enabled... Which is counter intuitive and generally should not be the case.

**Comment by t-wojciech** [comment]
Here are the new results from **[web frameworks benchmarker](https://github.com/the-benchmarker/web-frameworks)**.

> Pos | Language | Framework | Speed (64) | Speed (256) | Speed (512)
> 189 | r (4.0) | restrserve (0.3) | 2 754 | 2 652 | 2 582
> 192 | r (4.0) | httpuv (1.5) | 1 912 | 1 814 | 1 736
> 200 | r (4.0) | plumber (1.0) | 429 | 438 | 432

CC @vikram-rawat

**Pull request merged**